### PR TITLE
GPT training: use salmon DDP (+small refactors)

### PR DESCRIPTION
**Pull request by iejMac**

### CODE REVIEW

**Comment by iejMac** [comment]
something still seems wrong:

Screenshot 2024-09-21 at 7 47 07 PM

Screenshot 2024-09-21 at 7 47 20 PM

**Comment by iejMac** [comment]
issue was torch.compile, the DDP class needs to be made robust to that, more info on that at the bottom of https://pytorch.org/docs/main/notes/ddp.html#internal-design

after turning off torch.compile the situation looks ike this:
Screenshot 2024-09-22 at 1 07 05 AM
Screenshot 2024-09-22 at 1 07 20 AM

much closer perf and almost identical loss
